### DISCUSSION TITLE
Is Backlog within Json command supported?

### DISCUSSION THREAD

**Discussion by caenrigen** [arendst/Tasmota#18358] [Q&A] [answered]
Hey folks!

First of all thank you for the very nice framework! 😄

I think I am running into an unsupported feature or some kind of bug. Any help is appreciated.

I am in the Tasmota console. This works well:

```
01:18:19.495 CMD: Rule1 ON Clock#Timer=1 DO Backlog Color #FF000000D0; Wakeup 100 ENDON
01:18:19.501 RUL: Stored uncompressed, would compress from 63 to 57 (-10%)
01:18:19.505 RSL: RESULT = {"Rule1":{"State":"ON","Once":"OFF","StopOnError":"OFF","Length":63,"Free":448,"Rules":"ON Clock#Timer=1 DO Backlog Color #FF000000D0; Wakeup 100 ENDON"}}
```

but this does not work well and does something strange:

```
01:19:30.444 CMD: Json {"Rule1": "ON Clock#Timer=1 DO Backlog Color #FF000000D0; Wakeup 100 ENDON"}
01:19:30.630 RUL: Stored uncompressed, would compress from 45 to 39 (-14%)
01:19:30.634 RSL: RESULT = {"Rule1":{"State":"ON","Once":"OFF","StopOnError":"OFF","Length":45,"Free":466,"Rules":"ON Clock#Timer=1 DO Backlog Color #FF000000D0"}}
01:19:30.643 RSL: RESULT = {"Wakeup":"Started"}
01:19:30.692 RSL: RESULT = {"Wakeup":"Done","POWER":"ON","Dimmer":100,"Color":"FF000000FF","HSBColor":"0,100,100","White":100,"CT":500,"Channel":[100,0,0,0,100],"Scheme":1,"Fade":"OFF","Speed":1,"LedTable":"ON"}
01:19:59.303 CMD: Rule1
01:19:59.309 RSL: RESULT = {"Rule1":{"State":"ON","Once":"OFF","StopOnError":"OFF","Length":45,"Free":466,"Rules":"ON Clock#Timer=1 DO Backlog Color #FF000000D0"}}
```

Notice how the `"Rules"` are missing everything that was supposed to come after `;`. I am guessing this is somehow a problem with using `Backlog` within a `Json` command. 

Is this a bug?

---

PS I am running `Tasmota 12.0.2` on a smart lamp that came with it pre-installed.

**Comment by barbudor** [comment]
Why are you using json command instead of the straight command?

**Comment by caenrigen** [comment]
> Why are you using json command instead of the straight command?

I am sending a series of configurations to the device at once (sending them individually is a bit slow and more painful).

I know that I should be able to use Backlog command to achieve the same goal (did not try yet, I assume it works, does a Backlog within a Backlog work?), though Json is more convenient for my system. Besides the fact that I already have a bunch of "tools" developed that rely on the Json command (I am using Apple Shortcuts)

**Comment by barbudor** [comment]
Ok
So indeed there may be a problem with `;` inside a Json command.
Needs some investigation
May be Theo @arendst will have a quick solution as he signed the original implementation
Otherwise, I'll try to have a look this week end

**Comment by barbudor** [comment]
How are you sending the Json command?
Console? MQTT? Http?

**Comment by caenrigen** [comment]
> Ok
> So indeed there may be a problem with `;` inside a Json command.
> Needs some investigation
> May be Theo @arendst will have a quick solution as he signed the original implementation
> Otherwise, I'll try to have a look this week end

Thank you :)

> How are you sending the Json command?
> Console? MQTT? Http?

My full workflow is intended to use Http, but the behaviour seems the same when using the Console

**Comment by arendst** [comment, marked as answer]
It's not so much if backlog is supported within a json command as well using a rule command within a backlog command.

The json command is a wrapper around backlog commands. Backlog commands are terminated with ";" so executing a rule within a backlog command won't work.

You would see the same issue if you executed your rule within a backlog command.

Alas, not supported.

**Comment by sfromis** [comment]
One ugly workaround could be to first ensure that some variable contains the semicolon, obviously not via the JSON command.
`Mem9 ;`

And then:
`Json {"Rule1": "ON Clock#Timer=1 DO Backlog Color #FF000000D0%Mem9% Wakeup 100 ENDON"}`

Not really recommended, but not impossible either :grimacing:

**Comment by arendst** [comment]
Let me think about a solution where the json command only contains one command. 

This wouldn't need to be serviced by backlog but could be fed back to the command server.....

**Comment by arendst** [comment]
Try latest dev branch. It prioritzes compound commands like ``rule1`` and keeps non-compounded commands in backlog.

Solves your rule issue but introduces possible command sequence errors (although with rules commands I see no issue).

**Reply by caenrigen** [reply]
Thank you very much for looking into this!

I never flashed my device, so will take a bit if time to try it out. Let you know when I do

**Comment by anishsane** [comment]
I have a similar question about the websend command.

e.g. I can do this from a console or from console:

    backlog rule1 0; power1 1; rule1 1

I can also do it from HTTP (/cm?cmnd=) if I escape the special characters.

But same done via websend from another tasmota does not work:  

    websend [tasmota1.ip.address] backlog rule1 0; power1 1; rule1 1

It only executes `backlog rule1 0` on `tasmota1.ip.address`.

I can use the websend with the escaped URL like this though: 

    websend [tasmota1.ip.address] /cm?cmnd=backlog%20rule1%200%3b%20power1%2013b%20rule1%201

So, I wanted to check if it is unsupported by the design or if it is unsupported because it is uncommon.

**Reply by sfromis** [reply]
For me a test with websend+backlog directly from the web console worked without issues, with the expected commands arriving on the other end. Having `Weblog 3` active both ends should provide details about what happens where and when.
